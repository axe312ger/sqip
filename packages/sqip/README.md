# `sqip`

> TODO: description

## Usage

```
import sqip from 'sqip';

// TODO: DEMONSTRATE API
```
