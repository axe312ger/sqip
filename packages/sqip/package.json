{
  "name": "sqip",
  "version": "1.0.0-alpha.32",
  "description": "> TODO: description",
  "author": "Benedikt RÃ¶tsch <info@benedikt-roetsch.de>",
  "homepage": "https://github.com/axe312ger/sqip#readme",
  "license": "MIT",
  "scripts": {
    "build": "tsc -b",
    "watch": "tsc -w",
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "engines": {
    "node": ">=12.13.0"
  },
  "types": "dist/sqip.d.ts",
  "main": "dist/sqip.js",
  "directories": {
    "lib": "dist",
    "test": "__tests__"
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/axe312ger/sqip.git"
  },
  "bugs": {
    "url": "https://github.com/axe312ger/sqip/issues"
  },
  "keywords": [
    "lqip",
    "svg",
    "images",
    "placeholder",
    "webperf",
    "performance",
    "ux",
    "perceived"
  ],
  "dependencies": {
    "chalk": "^4.1.0",
    "cheerio": "^1.0.0-rc.3",
    "cli-table3": "^0.6.0",
    "expand-tilde": "^2.0.2",
    "fast-glob": "^3.2.4",
    "fs-extra": "^9.0.1",
    "node-vibrant": "^3.2.1-alpha.1",
    "probe-image-size": "^7.1.0",
    "sharp": "^0.28.1",
    "term-img": "^5.0.0"
  },
  "peerDependencies": {
    "sqip-plugin-blur": "^1.0.0-alpha.0",
    "sqip-plugin-data-uri": "^1.0.0-alpha.0",
    "sqip-plugin-primitive": "^1.0.0-alpha.0",
    "sqip-plugin-svgo": "^1.0.0-alpha.0"
  },
  "devDependencies": {
    "@types/debug": "^4.1.5",
    "@types/expand-tilde": "^2.0.0",
    "@types/fs-extra": "^9.0.11",
    "@types/probe-image-size": "^7.0.0",
    "debug": "^4.1.1"
  }
}
