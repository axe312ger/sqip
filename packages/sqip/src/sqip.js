// #############################################################################
// #
// # "SQIP" (pronounced \skwÉªb\ like the non-magical folk of magical descent)
// # is a SVG-based LQIP technique - https://github.com/technopagan/sqip
// #
// # Installation as CLI:
// # npm install -g sqip
// #
// # Installation as lib:
// # npm install sqip
// #
// # Requirements:
// # * Node.js >= v6 (https://nodejs.org/en/)
// #
// #############################################################################

import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'
import imageSize from 'probe-image-size'
import Vibrant from 'node-vibrant'
import sharp from 'sharp'
import termimg from 'term-img'
import Table from 'cli-table3'
import chalk from 'chalk'

import { locateFiles } from './helpers'

const debug = Debug('sqip')

const mainKeys = ['originalWidth', 'originalHeight', 'width', 'height', 'type']

const paletteKeys = [
  'Vibrant',
  'DarkVibrant',
  'LightVibrant',
  'Muted',
  'DarkMuted',
  'LightMuted'
]

// Resolves plugins based on a given config
// Array of plugin names or config objects, even mixed.
export async function resolvePlugins(plugins) {
  return Promise.all(
    plugins.map(async (plugin) => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin: Plugin.default }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )
}

async function processFile({ buffer, outputFileName, config }) {
  const { output, silent, parseableOutput } = config
  const result = await processImage({ buffer, config })
  const { content, metadata } = result
  let outputPath

  debug(`Processed ${outputFileName}`)

  // Write result svg if desired
  if (output) {
    try {
      // Test if output path already exists
      const stats = await fs.stat(output)

      // Throw if it is a file and already exists
      if (!stats.isDirectory()) {
        throw new Error(
          `File ${output} already exists. Overwriting is not yet supported.`
        )
      }
      outputPath = path.resolve(output, `${outputFileName}.svg`)
    } catch (err) {
      // Output directory or file does not exist. We will create it later on.
      outputPath = output
    }

    debug(`Writing ${outputPath}`)
    await fs.writeFile(outputPath, content)
  }

  // Gather CLI output information
  if (!silent) {
    if (outputPath) {
      console.log(`Stored at: ${outputPath}`)
    }

    // Generate preview
    if (!parseableOutput) {
      // Convert to png for image preview
      const preview = await sharp(Buffer.from(content)).png().toBuffer()

      try {
        termimg(preview, {
          fallback: () => {
            // SVG results can still be outputted as string
            if (metadata.type === 'svg') {
              console.log(content.toString())
              return
            }

            // No fallback preview solution yet for non-svg files.
            console.log(
              `Unable to render a preview for ${metadata.type} files on this machine. Try using https://iterm2.com/`
            )
          }
        })
      } catch (err) {
        if (err.name !== 'UnsupportedTerminalError') {
          throw err
        }
      }
    }

    // Metadata
    const tableConfig = parseableOutput && {
      chars: {
        top: '',
        'top-mid': '',
        'top-left': '',
        'top-right': '',
        bottom: '',
        'bottom-mid': '',
        'bottom-left': '',
        'bottom-right': '',
        left: '',
        'left-mid': '',
        mid: '',
        'mid-mid': '',
        right: '',
        'right-mid': '',
        middle: ' '
      },
      style: { 'padding-left': 0, 'padding-right': 0 }
    }

    // Figure out which metadata keys to show
    const allKeys = [...mainKeys, 'palette']
    const restMetadata = { ...metadata }
    allKeys.forEach((k) => delete restMetadata[k])

    const mainTable = new Table(tableConfig)
    mainTable.push(mainKeys)
    mainTable.push(mainKeys.map((key) => metadata[key]))
    console.log(mainTable.toString())

    // Show color palette
    const paletteTable = new Table(tableConfig)
    paletteTable.push(paletteKeys)
    paletteTable.push(
      paletteKeys
        .map((key) => metadata.palette[key].getHex())
        .map((hex) => chalk.hex(hex)(hex))
    )
    console.log(paletteTable.toString())

    Object.keys(restMetadata).forEach((key) => {
      console.log(chalk.bold(`${key}:`))
      console.log(restMetadata[key])
    })
  }

  return result
}

async function processImage({ buffer, config }) {
  const originalSizes = imageSize.sync(buffer)
  const vibrant = Vibrant.from(buffer)
  const palette = await vibrant.quality(0).getPalette()
  let metadata = {
    originalWidth: originalSizes.width,
    originalHeight: originalSizes.height,
    palette
  }

  // Load plugins
  const plugins = await resolvePlugins(config.plugins)

  // Interate through plugins and apply them to last returned image
  if (config.width > 0) {
    // Resize to desired output width
    buffer = await sharp(buffer).resize(config.width).toBuffer()

    const resizedMetadata = await sharp(buffer).metadata()
    metadata.width = resizedMetadata.width
    metadata.height = resizedMetadata.height
  } else {
    // Fall back to original size, keep image as is
    metadata.width = originalSizes.width
    metadata.height = originalSizes.height
  }

  for (const { name, options: pluginOptions, Plugin } of plugins) {
    debug(`Construct ${name}`)
    const plugin = new Plugin({
      sqipConfig: config,
      pluginOptions,
      metadata
    })
    debug(`Apply ${name}`)
    buffer = await plugin.apply(buffer)
  }

  return { content: buffer, metadata }
}

export default async function sqip(options) {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    width: 300,
    parseableOutput: false,
    silent: true
  }

  const config = Object.assign({}, defaultOptions, options)

  const { input, outputFileName, parseableOutput, silent } = config

  if (parseableOutput) {
    chalk.enabled = false
  }

  // Validate configuration
  if (!input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  // If input is a Buffer
  if (Buffer.isBuffer(input)) {
    return processFile({
      buffer: input,
      outputFileName,
      config
    })
  }

  const files = await locateFiles(input)
  debug('Found files:')
  debug(files)

  // Test if all files are accessable
  for (const file of files) {
    try {
      debug('check file ' + file)
      await fs.access(file, fs.constants.R_OK)
    } catch (err) {
      throw new Error(`Unable to read file ${file}`)
    }
  }

  // Iterate over all files
  const results = []
  for (const filePath of files) {
    // Apply plugins to files
    if (!silent) {
      console.log(`Processing: ${filePath}`)
    } else {
      debug(`Processing ${filePath}`)
    }
    const buffer = await fs.readFile(filePath)
    const result = await processFile({
      buffer,
      outputFileName: outputFileName || path.parse(filePath).name,
      config
    })

    results.push(result)
  }

  debug(`Finished`)

  // Return as array when input was array or results is only one file
  if (Array.isArray(input) || results.length === 0) {
    return results
  }
  return results[0]
}

export class SqipPlugin {
  constructor({ sqipConfig, metadata }) {
    this.sqipConfig = sqipConfig || {}
    this.metadata = metadata || {}
  }
}

export * from './helpers'
