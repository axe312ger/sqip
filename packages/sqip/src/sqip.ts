import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'
import imageSize from 'probe-image-size'
import Vibrant from 'node-vibrant'
import sharp from 'sharp'
import termimg from 'term-img'
import Table from 'cli-table3'
import chalk from 'chalk'

import { Palette } from '@vibrant/color'
import { OptionDefinition } from 'command-line-args'

import { locateFiles } from './helpers'

export { loadSVG, parseColor } from './helpers'

const debug = Debug('sqip')

const mainKeys = ['originalWidth', 'originalHeight', 'width', 'height', 'type']

const PALETTE_KEYS: (keyof Palette)[] = [
  'Vibrant',
  'DarkVibrant',
  'LightVibrant',
  'Muted',
  'DarkMuted',
  'LightMuted'
]

interface SqipPluginInterface {
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer
}

export interface SqipResult {
  content: Buffer
  metadata: SqipImageMetadata
}

export interface SqipPluginOptions {
  pluginOptions: PluginOptions
  options: PluginOptions
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
}
export interface SqipCliOptionDefinition extends OptionDefinition {
  description?: string
  required?: boolean
  default?: boolean
}

export interface PluginOptions {
  [key: string]: unknown
}

interface PluginResolver {
  name: string
  options?: PluginOptions
}

interface SqipConfig {
  input: string | Buffer
  outputFileName?: string
  output: string
  silent?: boolean
  parseableOutput?: boolean
  plugins: PluginType[]
  width?: number
}

interface ProcessFileOptions {
  buffer: Buffer
  outputFileName: string
  config: SqipConfig
}

interface SqipImageMetadata {
  originalWidth: number
  originalHeight: number
  palette: Palette
  height: number
  width: number
  type: string
  [key: string]: unknown
}

type PluginType = PluginResolver | string

export class SqipPlugin implements SqipPluginInterface {
  public metadata: SqipImageMetadata
  public sqipConfig: SqipConfig
  public options: PluginOptions
  static cliOptions: SqipCliOptionDefinition[]

  constructor(options: SqipPluginOptions) {
    const { sqipConfig, metadata } = options
    this.sqipConfig = sqipConfig || {}
    this.metadata = metadata || {}
    this.options = {}
  }
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer {
    throw new Error('Not implemented')
    return imageBuffer
  }
}

// Resolves plugins based on a given config
// Array of plugin names or config objects, even mixed.
export async function resolvePlugins(
  plugins: PluginType[]
): Promise<(PluginResolver & { Plugin: typeof SqipPlugin })[]> {
  return Promise.all(
    plugins.map(async (plugin) => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin: Plugin.default }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )
}

async function processFile({
  buffer,
  outputFileName,
  config
}: ProcessFileOptions) {
  const { output, silent, parseableOutput } = config
  const result = await processImage({ buffer, config })
  const { content, metadata } = result
  let outputPath

  debug(`Processed ${outputFileName}`)

  // Write result svg if desired
  if (output) {
    try {
      // Test if output path already exists
      const stats = await fs.stat(output)

      // Throw if it is a file and already exists
      if (!stats.isDirectory()) {
        throw new Error(
          `File ${output} already exists. Overwriting is not yet supported.`
        )
      }
      outputPath = path.resolve(output, `${outputFileName}.svg`)
    } catch (err) {
      // Output directory or file does not exist. We will create it later on.
      outputPath = output
    }

    debug(`Writing ${outputPath}`)
    await fs.writeFile(outputPath, content)
  }

  // Gather CLI output information
  if (!silent) {
    if (outputPath) {
      console.log(`Stored at: ${outputPath}`)
    }

    // Generate preview
    if (!parseableOutput) {
      // Convert to png for image preview
      const preview = await sharp(Buffer.from(content)).png().toBuffer()

      try {
        termimg(preview, {
          fallback: () => {
            // SVG results can still be outputted as string
            if (metadata.type === 'svg') {
              console.log(content.toString())
              return
            }

            // No fallback preview solution yet for non-svg files.
            console.log(
              `Unable to render a preview for ${metadata.type} files on this machine. Try using https://iterm2.com/`
            )
          }
        })
      } catch (err) {
        if (err.name !== 'UnsupportedTerminalError') {
          throw err
        }
      }
    }

    // Metadata
    const tableConfig = parseableOutput
      ? {
          chars: {
            top: '',
            'top-mid': '',
            'top-left': '',
            'top-right': '',
            bottom: '',
            'bottom-mid': '',
            'bottom-left': '',
            'bottom-right': '',
            left: '',
            'left-mid': '',
            mid: '',
            'mid-mid': '',
            right: '',
            'right-mid': '',
            middle: ' '
          },
          style: { 'padding-left': 0, 'padding-right': 0 }
        }
      : undefined

    // Figure out which metadata keys to show
    // @todo why is this unused?
    // const allKeys = [...mainKeys, 'palette']

    const mainTable = new Table(tableConfig)
    mainTable.push(mainKeys)
    mainTable.push(
      mainKeys.map((key) => String(metadata[key]) || 'can not display')
    )
    console.log(mainTable.toString())

    // Show color palette
    const paletteTable = new Table(tableConfig)
    paletteTable.push(PALETTE_KEYS)
    paletteTable.push(
      PALETTE_KEYS.map((key) => metadata.palette[key]?.hex)
        .filter<string>((hex): hex is string => typeof hex === 'string')
        .map((hex) => chalk.hex(hex)(hex))
    )
    console.log(paletteTable.toString())

    Object.keys(metadata)
      .filter((key) => ![...mainKeys, 'palette'].includes(key))
      .forEach((key) => {
        console.log(chalk.bold(`${key}:`))
        console.log(metadata[key])
      })
  }

  return result
}

interface ProcessImageOptions {
  buffer: Buffer
  config: SqipConfig
}

async function processImage({
  buffer,
  config
}: ProcessImageOptions): Promise<SqipResult> {
  const originalSizes = imageSize.sync(buffer)
  const vibrant = Vibrant.from(buffer)
  const palette = await vibrant.quality(0).getPalette()

  if (!originalSizes) {
    throw new Error('Unable to get image size')
  }

  const metadata: SqipImageMetadata = {
    originalWidth: originalSizes.width,
    originalHeight: originalSizes.height,
    palette,
    type: 'unknown',
    width: 0,
    height: 0
  }

  // Load plugins
  const plugins = await resolvePlugins(config.plugins)

  // Interate through plugins and apply them to last returned image
  if (config.width && config.width > 0) {
    // Resize to desired output width
    buffer = await sharp(buffer).resize(config.width).toBuffer()

    const resizedMetadata = await sharp(buffer).metadata()
    metadata.width = resizedMetadata.width || 0
    metadata.height = resizedMetadata.height || 0
  } else {
    // Fall back to original size, keep image as is
    metadata.width = originalSizes.width
    metadata.height = originalSizes.height
  }

  for (const { name, options: pluginOptions, Plugin } of plugins) {
    debug(`Construct ${name}`)
    const plugin = new Plugin({
      sqipConfig: config,
      pluginOptions: pluginOptions || {},
      metadata,
      options: {}
    })
    debug(`Apply ${name}`)
    buffer = await plugin.apply(buffer)
  }

  return { content: buffer, metadata }
}

export default async function sqip(
  options: SqipConfig
): Promise<SqipResult | SqipResult[]> {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    width: 300,
    parseableOutput: false,
    silent: true
  }

  const config = Object.assign({}, defaultOptions, options)

  const { input, outputFileName, parseableOutput, silent } = config

  if (parseableOutput) {
    chalk.level = 0
  }

  // Validate configuration
  if (!input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  // If input is a Buffer
  if (Buffer.isBuffer(input)) {
    if (!outputFileName) {
      throw new Error(
        `${outputFileName} is required when passing image as buffer`
      )
    }
    return processFile({
      buffer: input,
      outputFileName,
      config
    })
  }

  const files = await locateFiles(input)
  debug('Found files:')
  debug(files)

  // Test if all files are accessable
  for (const file of files) {
    try {
      debug('check file ' + file)
      await fs.access(file, fs.constants.R_OK)
    } catch (err) {
      throw new Error(`Unable to read file ${file}`)
    }
  }

  // Iterate over all files
  const results = []
  for (const filePath of files) {
    // Apply plugins to files
    if (!silent) {
      console.log(`Processing: ${filePath}`)
    } else {
      debug(`Processing ${filePath}`)
    }
    const buffer = await fs.readFile(filePath)
    const result = await processFile({
      buffer,
      outputFileName: outputFileName || path.parse(filePath).name,
      config
    })

    results.push(result)
  }

  debug(`Finished`)

  // Return as array when input was array or results is only one file
  if (Array.isArray(input) || results.length === 0) {
    return results
  }
  return results[0]
}

export * from './helpers'
