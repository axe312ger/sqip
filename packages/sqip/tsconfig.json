{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "typeRoots": [
      "./typings",
      "./node_modules/@types",
      "../../node_modules/@types"
    ]
  },
  "exclude": ["node_modules", "dist", "__tests__"]
}
