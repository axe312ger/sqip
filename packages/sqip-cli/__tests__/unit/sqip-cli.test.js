import sqip from 'sqip'
import sqipCLI from '../../src/sqip-cli'

jest.mock('sqip', () => ({
  default: jest.fn(),
  resolvePlugins: jest.fn(plugins =>
    plugins.map(plugin => ({
      name: plugin,
      Plugin: { cliOptions: [] }
    }))
  )
}))

const logSpy = jest.spyOn(global.console, 'log').mockImplementation(() => {})
const errorSpy = jest
  .spyOn(global.console, 'error')
  .mockImplementation(() => {})

global.process.exit = jest.fn()

describe('sqip-plugin-cli', () => {
  afterEach(() => {
    logSpy.mockClear()
    errorSpy.mockClear()
    global.process.exit.mockClear()
  })

  afterAll(() => {
    logSpy.mockReset()
    errorSpy.mockReset()
    global.process.exit.mockReset()
  })

  it('no args: show help and require input option', async () => {
    await sqipCLI()

    expect(sqip.default).not.toHaveBeenCalled()
    expect(sqip.resolvePlugins.mock.calls).toMatchSnapshot('resolve')
    expect(logSpy.mock.calls).toMatchSnapshot('log')
    expect(errorSpy.mock.calls).toMatchSnapshot('error')
    expect(global.process.exit).toHaveBeenCalledWith(1)
  })

  it('--version: show version', async () => {
    process.argv = [null, null, '--version']

    await sqipCLI()

    expect(sqip.default).not.toHaveBeenCalled()
    expect(sqip.resolvePlugins.mock.calls).toMatchSnapshot('resolve')
    expect(logSpy.mock.calls).toMatchSnapshot('log')
    expect(errorSpy.mock.calls).toMatchSnapshot('error')
    expect(global.process.exit).toHaveBeenCalledWith(0)
  })

  it('--help: show help', async () => {
    process.argv = [null, null, '--help']

    await sqipCLI()

    expect(sqip.default).not.toHaveBeenCalled()
    expect(sqip.resolvePlugins.mock.calls).toMatchSnapshot('resolve')
    expect(logSpy.mock.calls).toMatchSnapshot('log')
    expect(errorSpy.mock.calls).toMatchSnapshot('error')
    expect(global.process.exit).toHaveBeenCalledWith(0)
  })
})
