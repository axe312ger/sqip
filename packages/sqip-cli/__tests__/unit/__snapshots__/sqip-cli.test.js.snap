// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqip-plugin-cli --help: show help: error 1`] = `Array []`;

exports[`sqip-plugin-cli --help: show help: log 1`] = `
Array [
  Array [
    "
[4m[1msqip CLI[22m[24m

  Usage: sqip --input [path]                                                    
                                                                                
  \\"SQIP\\" (pronounced skwÉªb like the non-magical folk of magical descent) is a   
  SVG-based LQIP technique - https://github.com/technopagan/sqip                

[4m[1mOptions[22m[24m

  [1m-h[22m, [1m--help[22m [4mstring[24m        Show help                                            
  [1m--version[22m [4mstring[24m         Show version number                                  
  [1m-p[22m, [1m--plugins[22m [4mstring[][24m   One or more plugins. E.g. \\"-p primitive blur\\"        
  [1m-i[22m, [1m--input[22m [4mstring[24m                                                            
  [1m-o[22m, [1m--output[22m [4mstring[24m      Define the path of the resulting file. By default    
                           SQIP will guess the output file name.                
  [1m-w[22m, [1m--width[22m [4mnumber[24m       Width of the resulting file. Negative values and 0   
                           will fall back to original image width.              
  [1m--silent[22m                 Supress all output                                   
  [1m--parseable-output[22m       Ensure the output is parseable. Will suppress the    
                           preview images and the table borders.                

[4m[1mExamples[22m[24m

  Output input.jpg image as SQIP                                                
  $ sqip --input /path/to/input.jpg                                             
                                                                                
  Save input.jpg as result.svg with 25 shapes and no blur                       
  $ sqip -i input.jpg -n 25 -b 0 -o result.svg                                  
",
  ],
]
`;

exports[`sqip-plugin-cli --help: show help: resolve 1`] = `
Array [
  Array [
    Array [
      "primitive",
      "blur",
      "svgo",
    ],
  ],
  Array [
    Array [
      "primitive",
      "blur",
      "svgo",
    ],
  ],
]
`;

exports[`sqip-plugin-cli --version: show version: error 1`] = `Array []`;

exports[`sqip-plugin-cli --version: show version: resolve 1`] = `
Array [
  Array [
    Array [
      "primitive",
      "blur",
      "svgo",
    ],
  ],
]
`;

exports[`sqip-plugin-cli no args: show help and require input option: error 1`] = `
Array [
  Array [
    "
Please provide the following arguments: input",
  ],
]
`;

exports[`sqip-plugin-cli no args: show help and require input option: log 1`] = `
Array [
  Array [
    "
[4m[1msqip CLI[22m[24m

  Usage: sqip --input [path]                                                    
                                                                                
  \\"SQIP\\" (pronounced skwÉªb like the non-magical folk of magical descent) is a   
  SVG-based LQIP technique - https://github.com/technopagan/sqip                

[4m[1mOptions[22m[24m

  [1m-h[22m, [1m--help[22m [4mstring[24m        Show help                                            
  [1m--version[22m [4mstring[24m         Show version number                                  
  [1m-p[22m, [1m--plugins[22m [4mstring[][24m   One or more plugins. E.g. \\"-p primitive blur\\"        
  [1m-i[22m, [1m--input[22m [4mstring[24m                                                            
  [1m-o[22m, [1m--output[22m [4mstring[24m      Define the path of the resulting file. By default    
                           SQIP will guess the output file name.                
  [1m-w[22m, [1m--width[22m [4mnumber[24m       Width of the resulting file. Negative values and 0   
                           will fall back to original image width.              
  [1m--silent[22m                 Supress all output                                   
  [1m--parseable-output[22m       Ensure the output is parseable. Will suppress the    
                           preview images and the table borders.                

[4m[1mExamples[22m[24m

  Output input.jpg image as SQIP                                                
  $ sqip --input /path/to/input.jpg                                             
                                                                                
  Save input.jpg as result.svg with 25 shapes and no blur                       
  $ sqip -i input.jpg -n 25 -b 0 -o result.svg                                  
",
  ],
]
`;

exports[`sqip-plugin-cli no args: show help and require input option: resolve 1`] = `
Array [
  Array [
    Array [
      "primitive",
      "blur",
      "svgo",
    ],
  ],
]
`;

exports[`sqip-plugin-cli passes all given args to library 1`] = `
Array [
  Array [
    Object {
      "input": "foo",
      "output": "bar",
      "parseableOutput": undefined,
      "plugins": Array [
        Object {
          "name": "primitive",
          "options": Object {},
        },
        Object {
          "name": "blur",
          "options": Object {},
        },
        Object {
          "name": "svgo",
          "options": Object {},
        },
      ],
      "silent": undefined,
      "width": 1337,
    },
  ],
]
`;
